### PR TITLE
ARROW-3721: [Gandiva] [Python] Support all Gandiva literals

### PR DESCRIPTION
**Pull request by pcmoritz**

### CODE REVIEW

File: python/pyarrow/gandiva.pyx [hunk 1]
```diff
@@ -150,10 +163,49 @@ cdef class Filter:
             batch.sp_batch.get()[0], selection))
         return SelectionVector.create(selection)
 
+
+cdef inline DataType _ensure_type(object type):
```
Comment (kszucs) [inline python/pyarrow/gandiva.pyx]: There is an `_as_type` function defined in `types.pxi`

File: python/pyarrow/gandiva.pyx [hunk 2]
```diff
@@ -150,10 +163,49 @@ cdef class Filter:
...
+        return type
+
+
 cdef class TreeExprBuilder:
 
-    def make_literal(self, value):
-        cdef shared_ptr[CNode] r = TreeExprBuilder_MakeLiteral(value)
+    def make_literal(self, value, dtype):
+        cdef shared_ptr[CNode] r
+        cdef DataType type = _ensure_type(dtype)
```
Comment (kszucs) [inline python/pyarrow/gandiva.pyx]: Tests for passing types as string and a failing case?

File: python/pyarrow/gandiva.pyx (hunk 1 above)
Comment (pcmoritz) [reply python/pyarrow/gandiva.pyx]: Thanks, I'm using that now (but it requires to make that symbol available in lib.pxd, not sure if we want that).

File: python/pyarrow/gandiva.pyx (hunk 2 above)
Comment (pcmoritz) [reply python/pyarrow/gandiva.pyx]: Added, thanks!

File: python/pyarrow/gandiva.pyx (hunk 1 above)
Comment (kszucs) [reply python/pyarrow/gandiva.pyx]: Converted it to a pure python function. We have another `cdef inline _ensure_type` in array.pxi which causes other issues, will resolve it in a follow-up PR, see https://issues.apache.org/jira/browse/ARROW-3735

File: python/pyarrow/gandiva.pyx (hunk 2 above)
Comment (kszucs) [reply python/pyarrow/gandiva.pyx]: Thanks!

**Review by kszucs — APPROVED** [review]
LGTM, merging on green

File: python/pyarrow/tests/test_gandiva.py [hunk 3]
```diff
@@ -91,10 +91,73 @@ def test_filter():
...
+    builder.make_literal(6, "int32")
+    builder.make_literal(7, "int64")
+    builder.make_literal(8.0, "float32")
+    builder.make_literal(9.0, "float64")
+    builder.make_literal("hello", "string")
+    builder.make_literal(b"world", "binary")
+
+    with pytest.raises(TypeError):
+        builder.make_literal("hello", pa.int64())
+        builder.make_literal(True, None)
```
Comment (kszucs) [inline python/pyarrow/tests/test_gandiva.py]: This line isn't running.

**Comment by pcmoritz** [comment]
Thanks @kszucs

**Pull request closed without merging**